[pt2][inductor]global autotuning cache

nmacchioni opened the pull request:
Summary:
this diff adds logic to handle a global autotuning cache, stored in json format at config.global_cache_path.

what is changing from `DiskCache`:
* `DiskCache` is renamed to `PersistentCache`
* the local cache is now stored as a single file in json format, located at `/tmp/torchinductor_{$USER}/local_cache`. the file contains a dictionary structure like `local_cache[name][inputs][choice]` where `name` is the type of operation, like `addmm`, `inputs` is the repr of the inputs, and `choice` is the hash of a `ChoiceCaller`. the stored value is the benchmark time for that `ChoiceCaller`.
* a global cache is added, initially stored at `fbcode/caffe2/torch/_inductor/global_cache`, with almost identical format as the local cache. since the global cache exists over different machines, there is an additional `dinfo` field, such that `global_cache[dinfo] = local_cache` (at least structure wise, there is no guarantee that the global cache and local cache share the same values). `dinfo` is just a repr of the cuda device properties.
* the autotuner will prioritize the global cache, and return values from there first, before looking in the local cache
* the autotuner will look in both the global cache and the local cache even when `max_autotune=False`, but will still only generate values if `max_autotune=True`.
* the autotuner will log global cache hits and misses to a scuba table (inductor_autotuning_cache) which will be used to update the global cache at regular intervals

Test Plan: D43285472

Differential Revision: D42785435



cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

jansel commented on torch/_inductor/codecache.py:
```diff
@@ -56,40 +66,99 @@ def _compile_end():
-    def lookup(self, key: Any, generate: Callable[[], Any]):
+    def update_local_cache(self, local_cache):
+        counters["inductor"]["select_algorithm_autotune_local_cache_updated"] += 1
+        write_atomic(self.local_cache_path, json.dumps(local_cache, indent=4))
```
I worry that this pattern of load/update/write will be `O(n^2)`, as the cache gets very big this will get more and more expensive.

Perhaps it is small enough it won't matter though.

nmacchioni replied:
I'd expect the local cache to be almost always empty, or very close to empty, otherwise we aren't doing a very good job of updating the global cache.

nmacchioni committed:
[pt2][inductor]global autotuning cache (#94922)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/94922

this diff adds logic to handle a global autotuning cache, stored in json format at config.global_cache_path.

what is changing from `DiskCache`:
* `DiskCache` is renamed to `PersistentCache`
* the local cache is now stored as a single file in json format, located at `/tmp/torchinductor_{$USER}/local_cache`. the file contains a dictionary structure like `local_cache[name][inputs][choice]` where `name` is the type of operation, like `addmm`, `inputs` is the repr of the inputs, and `choice` is the hash of a `ChoiceCaller`. the stored value is the benchmark time for that `ChoiceCaller`.
* a global cache is added, initially stored at `fbcode/caffe2/torch/_inductor/global_cache`, with almost identical format as the local cache. since the global cache exists over different machines and is persistent, there are additional fields `dinfo` and `vinfo`, such that `global_cache[dinfo][vinfo] = local_cache` (at least structure wise, there is no guarantee that the global cache and local cache share the same values). `dinfo` is a repr of the cuda device properties and `vinfo` is a repr of the cuda version.
* the autotuner will prioritize the global cache, and return values from there first, before looking in the local cache
* the autotuner will look in both the global cache and the local cache even when `max_autotune=False`, so long as `search_autotune_cache=True`, but will still only generate values if `max_autotune=True`.
* the autotuner will log global cache hits and misses to a scuba table (inductor_autotuning_cache) which will be used to update the global cache at regular intervals

Test Plan:
`buck test mode/opt caffe2/test/inductor:select_algorithm`

custom unit tests coming in D43285472

Reviewed By: jansel, jiawenliu64

Differential Revision: D42785435

fbshipit-source-id: 64bc23f4ce098d7c29856166dcd23661c4fd2431
